service/dap/server.go

```diff
@@ -451,6 +451,44 @@ func (s *Server) onContinueRequest(request *dap.ContinueRequest) {
 	s.doContinue()
 }
 
+func (s *Server) onThreadsRequest(request *dap.ThreadsRequest) {
+	if s.debugger == nil {
+		s.sendErrorResponse(request.Request, UnableToDisplayThreads, "Unable to display threads", "debugger is nil")
+		return
+	}
+	gs, _, err := s.debugger.Goroutines(0, 0)
+	if err != nil {
+		switch err.(type) {
+		case *proc.ErrProcessExited:
+			// If the program exits very quickly, the initial threads request will complete after it has exited.
+			// A TerminatedEvent has already been sent. Ignore the err returned in this case.
+			s.send(&dap.ThreadsResponse{Response: *newResponse(request.Request)})
+		default:
+			s.sendErrorResponse(request.Request, UnableToDisplayThreads, "Unable to display threads", err.Error())
+		}
+		return
+	}
+
+	threads := make([]dap.Thread, len(gs))
+	if len(threads) == 0 {
+		threads = []dap.Thread{{Id: 1, Name: "Dummy"}}
```

Comment:
(out of curiosity) why is this dummy one necessary?

Reply:
I'm ok with the PR as is, but one way of dealing with this and also with the fact that there's no distinction between threads and  goroutines is to loop through debugger.Threads() and add every thread that has `GoroutineID == 0` and set `dap.Thread.Id = -thread.ID` (i.e. os threads have negative IDs, goroutines have positive IDs).

There's always going to be at least one OS thread so the dummy response wouldn't be needed anymore (however this would mean that negative thread IDs have to be handled throughout the API).

Reply:
For now my goal was to keep feature parity with the existing implementation, but we can definitely improve things going forward. I can add this to my personal TODO list or we can file an issue in this repo to investigate this further.

Reply:
I was about to file an issue for this, but have a follow-up question. A StackTrace request will be issued for each reported thread. For the dummy thread, we end up with an unknown goroutine error, so the editor displays "unable to provide stack trace" for it. What would be the behavior for the stacktrace behavior for OS threads?

Reply:
I'm not sure I understand the question. If you're asking how the stacktrace api interprets its GoroutineID parameter then you can look at proc.FindGoroutine in pkg/proc/variables.go. Basically:

* `GoroutineID == -1` the currently selected goroutine or the current thread if it isn't running a goroutine
* `GoroutineID == 0` the current thread if it isn't running a goroutine (an error otherwise)
* `GoroutineID > 0` the specified goroutine if it can be found